### PR TITLE
Parametrize buried soft fork in regtest and refactor

### PR DESCRIPTION
**Pull request by NicolasDorier**
This PR has three goals:
- Parametrizes buried soft fork in regtest (Same approach than @sdaftuar on  https://github.com/bitcoin/bitcoin/pull/8418)
- Refactor by using a vector for buried deployment. Buried deployments will have its own BIP (https://github.com/bitcoin/bitcoin/pull/8391#issuecomment-237584871) so it makes sense to have a consistent way of adding a new one which can be parametrized out of the box by 297a65728de5fdd843335c8c9fc3821ae5d5bac4.
- Introduce CChainParams::AllowsOverriddenSoftFork so we don't abuse MineBlocksOnDemand for bip9 and buried sf overrides.

@sipa, you asked for it on https://github.com/bitcoin/bitcoin/pull/8391#issuecomment-236003190.
@jtimon, using similar code you used on https://github.com/jtimon/bitcoin/compare/remove-ism...consensus-post-remove-ism

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -316,6 +316,24 @@ const CChainParams &Params() {
...
+        }
+        if (deployment == Consensus::BIP65_HEIGHT_ACTIVE) {
+                consensus.BIP65Height = nStartHeight;
+        }
+        if (deployment == Consensus::BIP66_HEIGHT_ACTIVE) {
+                consensus.BIP66Height = nStartHeight;
+        }
+}
+
+void UpdateRegtestBuriedDeploymentParameters(Consensus::BuriedDeploymentPos deployment, int64_t nStartHeight)
```
Comment (jtimon) [inline src/chainparams.cpp]: More generic to take the the chainparams petname. Maybe assert that is equal to CBaseChainParams::REGTEST.

File: src/init.cpp [hunk 2]
```diff
@@ -1026,6 +1027,35 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
         }
     }
 
+    if (!mapMultiArgs["-buriedsfparams"].empty()) {
+        // Allow overriding ism parameters for testing
+        if (!Params().MineBlocksOnDemand()) {
```
Comment (jtimon) [inline src/init.cpp]: Instead of abusing MineBlocksOnDemand(), maybe a new attribute in CChainParams?
What about fAllowOverwriteBuriedSoftforks?
We don't even need a new getter, CChainParams is accessed as a const ref everywhere outside chainparams.cpp.

**Comment by jtimon** [comment]
Concept ACK

File: src/chainparams.h [hunk 3]
```diff
@@ -101,6 +105,8 @@ class CChainParams
  */
 const CChainParams &Params();
 
+void UpdateRegtestBuriedDeploymentParameters(Consensus::BuriedDeploymentPos deployment, int64_t nStartHeight);
```
Comment (jtimon) [inline src/chainparams.h]: This is not necessary anymore.

File: src/chainparams.h [hunk 4]
```diff
@@ -56,7 +56,8 @@ class CChainParams
     const Consensus::Params& GetConsensus() const { return consensus; }
     const CMessageHeader::MessageStartChars& MessageStart() const { return pchMessageStart; }
     int GetDefaultPort() const { return nDefaultPort; }
-
+    /** This chain can have buried and bip9 soft fork parameters overridden with -bip9params and -buriedsfparams */
+    bool AllowsOverriddenSoftFork() const { return fAllowsOverriddenSoftFork; }
```
Comment (jtimon) [inline src/chainparams.h]: Is the getter necessary? Params() already returns a const ref, why not use Params().fAllowsOverriddenSoftFork directly in init (making fAllowsOverriddenSoftFork public)?

Comment (NicolasDorier) [inline src/chainparams.h]: I'm following the coding style of previous code in CChainParams which does not expose the fields.

**Comment by jtimon** [comment]
utACK 3502867

File: src/chainparams.h (hunk 4 above)
Comment (jtimon) [inline src/chainparams.h]: fair enough

File: qa/rpc-tests/bip65-cltv-p2p.py [hunk 5]
```diff
@@ -43,7 +43,7 @@ def __init__(self):
     def setup_network(self):
         # Must set the blockversion for this test
         self.nodes = start_nodes(self.num_nodes, self.options.tmpdir,
-                                 extra_args=[['-debug', '-whitelist=127.0.0.1', '-blockversion=3']],
+                                 extra_args=[['-debug', '-whitelist=127.0.0.1', '-blockversion=3', '-buriedsfparams=bip65:1351']],
```
Comment (MarcoFalke) [inline qa/rpc-tests/bip65-cltv-p2p.py]: Nit: Would it make sense to revert https://github.com/bitcoin/bitcoin/pull/8391/files#diff-30641a8ce712805f484b8305049f04b3 to the old code as well?

File: src/init.cpp [hunk 6]
```diff
@@ -993,9 +994,10 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 
     if (!mapMultiArgs["-bip9params"].empty()) {
         // Allow overriding BIP9 parameters for testing
-        if (!Params().MineBlocksOnDemand()) {
-            return InitError("BIP9 parameters may only be overridden on regtest.");
+        if (!Params().AllowsOverriddenSoftFork()) {
```
Comment (MarcoFalke) [inline src/init.cpp]: Nit: I think keeping this the way it was is just fine. (Note that `fAllowsOverriddenSoftFork` is not really a chain param, is it?)

Comment (jtimon) [inline src/init.cpp]: He is creating a new one. Maybe a future testnet allows this but does not mine blocks on demand. That's why we have different properties for chainparams instead of calling IsTestnet() or IsRegtest () like we did before.

File: qa/rpc-tests/bip65-cltv-p2p.py (hunk 5 above)
Comment (NicolasDorier) [inline qa/rpc-tests/bip65-cltv-p2p.py]: This particular change does not change anything in fact, because it is setting the buriefsf height at its default value anyway.
I can't really revert the rest. The reason is that the previous test was testing activation and enforcement. But now we don't have different height for activation and enforcement anymore.

File: src/init.cpp (hunk 6 above)
Comment (NicolasDorier) [inline src/init.cpp]: What do you mean by "fAllowsOverriddenSoftFork" ? where would you put it ?

File: qa/rpc-tests/bip65-cltv-p2p.py (hunk 5 above)
Comment (MarcoFalke) [inline qa/rpc-tests/bip65-cltv-p2p.py]: I have in mind you changed the heights for enforcement to be the same for all tests? So you could revert the patch that makes the test mine additional blocks and reduce `buriedsfparams` accordingly.

Comment (NicolasDorier) [inline qa/rpc-tests/bip65-cltv-p2p.py]: yes good point.

**Comment by NicolasDorier** [comment]
@MarcoFalke I reverted the python test height to old value, as well as deactivating the buried soft forks by default on regtest.

**Comment by sipa** [comment]
What is the status here?

**Comment by NicolasDorier** [comment]
I think it is ready, would appreciate some feedback. As suggested by @sdaftuar I did a small [BIP draft](https://docs.google.com/document/d/1vhE91A_GFJkB6BdoqcnaW5Q2zvnBbM2MPSwiL7h6LEc/edit?usp=sharing) for buried deployments.

File: src/init.cpp (hunk 2 above)
Comment (sipa) [inline src/init.cpp]: Agree with using a new attribute.

File: src/chainparams.h (hunk 3 above)
Comment (sipa) [inline src/chainparams.h]: What is not necessary anymore?

File: src/main.cpp [hunk 7]
```diff
@@ -2348,7 +2348,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     // before the first had been spent.  Since those coinbases are sufficiently buried its no longer possible to create further
     // duplicate transactions descending from the known pairs either.
     // If we're on the known chain at height greater than where BIP34 activated, we can save the db accesses needed for the BIP30 check.
-    CBlockIndex *pindexBIP34height = pindex->pprev->GetAncestor(chainparams.GetConsensus().BIP34Height);
+    CBlockIndex *pindexBIP34height = pindex->pprev->GetAncestor(chainparams.GetConsensus().vBuriedDeploymentHeights[Consensus::BIP34_HEIGHT_ACTIVE]);
     //Only continue to enforce if we're below BIP34 activation height or the block hash at that height doesn't correspond.
     fEnforceBIP30 = fEnforceBIP30 && (!pindexBIP34height || !(pindexBIP34height->GetBlockHash() == chainparams.GetConsensus().BIP34Hash));
```
Comment (sipa) [inline src/main.cpp]: If you make the bip34 height configurable, you should also make the hash configurable, or this test will fail.

File: src/init.cpp (hunk 6 above)
Comment (sipa) [inline src/init.cpp]: I'd say this is fine as a chain param.

File: src/main.cpp (hunk 7 above)
Comment (NicolasDorier) [inline src/main.cpp]: meh, on Regtest BIP34Hash has always been 0. Ability to override BIP34Hash should be done when at least one unit test will make use of it, which is not the case for now.

File: src/chainparams.cpp [hunk 8]
```diff
@@ -356,4 +348,4 @@ void UpdateRegtestBIP9Parameters(Consensus::DeploymentPos d, int64_t nStartTime,
 {
     regTestParams.UpdateBIP9Parameters(d, nStartTime, nTimeout);
 }
-
```
Comment (jtimon) [inline src/chainparams.cpp]: style nit: unnecessary diff.

File: src/init.cpp (hunk 6 above)
Comment (jtimon) [inline src/init.cpp]: nit s/Params()/chainparams/

File: src/init.cpp [hunk 9]
```diff
@@ -993,9 +994,10 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
+            return InitError("BIP9 parameters can't be overriden on this chain.");
         }
+        CChainParams& mutableParams = Params(Params().NetworkIDString());
```
Comment (jtimon) [inline src/init.cpp]: nit s/Params()/chainparams/

File: src/init.cpp [hunk 10]
```diff
@@ -1026,6 +1028,36 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
         }
     }
 
+    if (!mapMultiArgs["-buriedsfparams"].empty()) {
+        // Allow overriding ism parameters for testing
+        if (!Params().AllowsOverriddenSoftFork()) {
```
Comment (jtimon) [inline src/init.cpp]: nit s/Params()/chainparams/

File: src/init.cpp [hunk 11]
```diff
@@ -1026,6 +1028,36 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
+            return InitError("Buried soft fork parameters can't be overridden on this chain.");
+        }
+        CChainParams& mutableParams = Params(Params().NetworkIDString());
```
Comment (jtimon) [inline src/init.cpp]: nit s/Params()/chainparams/

**Comment by jtimon** [comment]
re-utACK besides nits.

**Comment by NicolasDorier** [comment]
nit addressed @MarcoFalke this PR might interest you as it touches the tests.

**Comment by NicolasDorier** [comment]
is it possible to run the tests again ? I don't think the failure is related to this PR.

**Comment by sipa** [comment]
@NicolasDorier I restarted the failing tests.

**Comment by MarcoFalke** [comment]
@NicolasDorier as I see correctly, the tests are no longer failing when there is an off-by-one error. What do you think of https://github.com/MarcoFalke/bitcoin/commits/Mf1609-qaReworkISM ?

**Comment by NicolasDorier** [comment]
@MarcoFalke good point, just cherry picked your commit.

**Comment by MarcoFalke** [comment]
Concept ACK 97c4507 (can't review my own changes, though)

**Review by jtimon — APPROVED** [review]

**Review by jtimon — APPROVED** [review]

**Comment by jtimon** [comment]
Mhmm, there is a new review feature...
utACK 43de0ce 
concept ACK 97c4507

**Comment by MarcoFalke** [comment]
This needs the merge conflict solved.

**Comment by NicolasDorier** [comment]
rebased

**Comment by NicolasDorier** [comment]
closing it, too much conflict, if someone is interested I might do the work again.
